firefox: add PGO support

S-NA opened the pull request:
This is a draft PR which hopefully will eventually be able to be merged to close issue #76484.

As it stands, I have not built these changes but assume they work as I tested something similar in the past. The biggest issue blocking this is having the hashes for the pgodata get automatically updated. Any insight is appreciated. I am still unsure if how I am going about doing this is even the best way. I have considered making separate packages entirely for the pgodata but exposing them at the top-level seems wrong but that seems to be the only way to update them using the `update-source-version`.

S-NA committed:
firefox: add PGO support

PGO is currently Linux 64-bit only.

Updating Firefox requires some manual assistance which is unwanted. At
the moment the PGO data is fetched from Mozilla's CI and needs to have
its hash updated manually.

The revision for the PGO data is automatically fetched and updated
which when doing a typical update of Firefox. However, it involves
downloading the Firefox source which is slow. It is also sort of a hack
in that, it passes through rev and update-source-version allows you to
specify a revision.

SuperSandro2000 commented on pkgs/applications/networking/browsers/firefox/common.nix:
```diff
@@ -269,6 +285,11 @@ buildStdenv.mkDerivation ({
     # 60.5+ & 66+ did split the google API key arguments: https://bugzilla.mozilla.org/show_bug.cgi?id=1531176
     configureFlagsArray+=("--with-google-location-service-api-keyfile=$TMPDIR/ga")
     configureFlagsArray+=("--with-google-safebrowsing-api-keyfile=$TMPDIR/ga")
+  '') + (lib.optionalString (ltoSupport && pgoSupport) ''
+    configureFlagsArray+=("--enable-lto=cross")
+    configureFlagsArray+=("--enable-profile-use=cross")
+    configureFlagsArray+=("--with-pgo-jarlog=${pgo}/en-US.log")
+    configureFlagsArray+=("--with-pgo-profile-path=${pgo}/merged.profdata")
   '') + ''
```
```suggestion
  '') + lib.optionalString (ltoSupport && pgoSupport) ''
    configureFlagsArray+=("--enable-lto=cross" "--enable-profile-use=cross" "--with-pgo-jarlog=${pgo}/en-US.log" "--with-pgo-profile-path=${pgo}/merged.profdata")
  '' + ''
```

mweinelt replied:
This breaks readability big time, please don't.

SuperSandro2000 replied:
```suggestion
  '') + lib.optionalString (ltoSupport && pgoSupport) ''
    configureFlagsArray+=(
      "--enable-lto=cross"
      "--enable-profile-use=cross"
      "--with-pgo-jarlog=${pgo}/en-US.log"
      "--with-pgo-profile-path=${pgo}/merged.profdata"
    )
  '' + ''
```

Atemu replied:
An even better solution is to just add those flags to configureFlags: https://github.com/Atemu/nixpkgs/commit/0d94218207d76299d5f44d41bdcd09007d86ac77.

Atemu commented:
I tried it and it doesn't seem to make a difference at all. Speedometer is still around 90 while it's 118 in firefox-bin.

Something isn't right here.

Atemu commented:
The [Gentoo ebuild](https://gitweb.gentoo.org/repo/gentoo.git/tree/www-client/firefox/firefox-89.0.2.ebuild) adds `MOZ_PGO=1` which we don't; that sounds important. Rebuilding...

Atemu commented:
That's not it either :/

S-NA commented:
> Speedometer is still around 90

Does that mean it also was measuring 90 before building with PGO? If that is the case, then yes something is really wrong. I will try to get around testing these changes to see what is wrong. The only other thing I can think of, that I should have done was build with `++ lib.optional (!buildStdenv.isx86_32 && !buildStdenv.Aarch32 && !buildStdenv.Aarch64) "--enable-rust-simd"` since upstream does that. However, if you are seeing no improvement over firefox without PGO then that is alarming.

S-NA committed:
rust: reintroduce version 1.51

Mozilla's CI currently uses LLVM 11 and rustc 1.51. To be able to use
their PGO data, Firefox must be built with them too otherwise the PGO
data becomes invalidated (due to the actual code generation differing
from what is expected) and little to no optimization can be done.

S-NA commented:
So it seems that because Mozilla's CI uses LLVM 11 and rustc 1.51 the PGO data becomes stale when applying it to a build done with LLVM 12 and rustc 1.53. I think more than anything is the newer LLVM version causes code generation to differ greatly invalidating a lot of the data gathered.

The only solution is to match Mozilla's CI compiler versions. Which is what I do with reintroducing rustc 1.51. On a good note firefox-esr seems fine using the compiler version it is at currently so there is no need for another rustc for it, but I am not sure if it is wanted to lug around an older version of rustc just to have Firefox PGO. For me it shows an ~15% improvement on Speedometer, whether this is worth the additional rustc in nixpkgs is really an open question.

Atemu commented:
Currently compiling your newest head (LLVM11, Rustc 1.51) but I'm still seeing lots of `Function control flow change detected (hash mismatch)` in the log.

I don't think there's too much overhead in keeping a slightly older version of rustc around that sits between two we need to support anyways or ist rustc prone to breaking often?

Atemu commented:
It works! Now getting consistent 124 points instead of 98!

mweinelt commented on pkgs/applications/networking/browsers/firefox/common.nix:
```diff
@@ -91,13 +93,17 @@ let
 
   # 78 ESR won't build with rustc 1.47
-  inherit (if lib.versionAtLeast ffversion "82" then rustPackages else rustPackages_1_45)
+  inherit (if lib.versionAtLeast ffversion "82" then rustPackages_1_51 else rustPackages_1_45)
```
Please add a comment how to discover the rust version Mozilla uses for their stable firefox builds.

mweinelt commented:
Works for me as well. From 115 to 139 on an empty profile. Chromium is at 130 btw. So this is awesome. Could/should we add a changelog entry for that?

ajs124 commented:
Sorry if this has been discussed somewhere and I've just not seen it, but in case it hasn't:
Why aren't we just doing our own profiling run and generating the data ourselves during the build? That way we'd be able to use whichever LLVM and rustc version we'd want. I seem to remember other distributions doing it that way.
I understand that it would probably significantly increase the build time, but personally, I'd rather have a longer build than having to maintain a separate rust version and have to stick with the LLVM version mozilla happens to use.

Atemu commented:
The main problem with that is r13y; it's in no way deterministic.

Since we need pre-generated profiles for deterministic builds, we might as well use Mozilla's profiles and they work very well it seems.

mweinelt commented:
Quite plainly, this increases the workload of the people that backport this into stable releases.

We are already carrying a custom version of `rust-cbindgen` and `nss` into stable releases, and now we are supposed to add a custom `rustc` version into the mix. :disappointed:

ajs124 commented:
I wouldn't exactly call "requires another FOD and a bespoke LLVM and rustc version, which we have to maintain separately, all while still throwing warnings"  "working well".

Plus I highly doubt firefox builds reproducably as is.

mweinelt commented:
```
Mozilla Firefox 90.0
derivation '/nix/store/4mr7clyq2hb0lsj1vw3bx960yw26bk09-firefox-unwrapped-90.0.drv' may not be deterministic: output '/nix/store/qlgyx58rjzk0z7milxw8010m34b2ixlw-firefox-unwrapped-90.0' differs
error: build of '/nix/store/4mr7clyq2hb0lsj1vw3bx960yw26bk09-firefox-unwrapped-90.0.drv' failed
```

FWIW: Firefox is not reproducible on x86_64 right now.

S-NA commented:
> Currently compiling your newest head (LLVM11, Rustc 1.51) but I'm still seeing lots of `Function control flow change detected (hash mismatch)` in the log.

Between the two build logs I had, there was a lot more of such warnings when using the newer compiler versions. I suspect the current warnings are from different build options, and since those are not tied to anything performance critical they do not really hurt. They can silenced with a diagnostic related flag to Clang but that really does not need to be done. 


> Why aren't we just doing our own profiling run and generating the data ourselves during the build?

We could do this but as others have mentioned it costs the ability to be deterministic. With doing the profiling twice on the same machine, I could produce different profiles. With how we are now, determinism is mostly limited to the compiler failing us. https://bugzilla.mozilla.org/show_bug.cgi?id=885777 In the future things should get better, either as Clang gets better with code generation or upstream decides to work around non-deterministic code generation. With doing our own profiling we will never be able to be deterministic purely from how sensitive profiling is.

However, I am fine trying to implement PGO doing our own profiling and generating if that is what the consensus wants.

mweinelt commented:
Using the binary pgo data takes us further away from the source build we are trying to achieve here. If people want the quick and easy solution they can already use `firefox-bin` instead. Let's please keep this difference intact.

I'm strongly pro generating the pgo data ourselves if possible.

ashkitten commented:
as a compromise, maybe we can pregenerate pgo data as part of an update script? that way it's still reproducible but we're not dependent on mozilla for it and can use up-to-date versions of rust and llvm

dasJ commented:
I wouldn't call the PGO data unfree or a blob, just binary. While it is a binary format, so is PNG and we already have a whole lot of packages that have them in them.

The PGO format is llvm-specific and not something that is Mozilla-proprietary. Downloading the [PGO data](https://firefox-ci-tc.services.mozilla.com/api/index/v1/task/gecko.v2.mozilla-release.revision.9fcea995d1dabc5a4f4ef3811dc0e6e00d88cbe3.firefox.linux64-profile/artifacts/public/build/profdata.tar.xz) gives us a tar with profdata file in it.
`llvm-profdata show -all-functions merged.profdata` (with llvm13, may also work with an older version) gives us the data that is stored in the file - it's not arbitrary bytecode that we are used from other blobs like firmware blobs.

Generating the data ourselves would **really** push r13y of the nix sandbox to the limits. While the current Firefox is not reproducible afaik, this would most certainly make it an absolute nightmare.
Mozilla's PGO data will probably give us more useful data than just "we started the browser and quit it". Even if they just give us that and no "real" browsing where somewhat user activity is sampled, we still get a fully reproducible and downloadable file that we can incorperate into our build without going through the valley of pain that is spawning an entire graphical environment and generating the file.

mweinelt commented:
I do feel these changes are indeed worthwhile and should be rebased once more so we can include them. I've changed my mind over my comment in https://github.com/NixOS/nixpkgs/pull/129503#issuecomment-879178093, given that we probably won't ever get our own meaningful PGO data. And not having PGO when the work has been done right here is disappointing for many users (me included) I'd imagine.

Atemu commented:
While I don't agree that PGO profiles are on the same level as PNGs (the compiler uses them to make decisions!):

A. it only influences compiler decisions insofar as *how* to achieve the task, not *what* task. (The end result will do the same thing but with different performance characteristics)
B. it's inherently non-reproducible; the data depends on the behaviour of a *running* system as well as the regular sources. Mozilla's data isn't reproducible in any way either.
C. it's hard to generate this ourselves (needs a representative set of websites and tasks to perform while profiling and representative hardware)

Things to clear up on this:

* Does/could Mozilla (or another trusted entity) generate PGO data for multiple rust versions? We can't be the only ones building a stable distro that needs PGO data for older rust versions.
* How much overhead is it to keep rust up do date in stable and is it something that might need to happen anyways?
* How does Mozilla generate their PGO profiles?
* If generating our own PGO data in a drv isn't possible, couldn't we generate PGO data impurely and add it as a hashed BLOB?

On the topic of generating PGO data: Wouldn't it require a two-staged build where the regular, unoptimised Firefox is used for profiling and then a new optimised one is built with that data?

mweinelt commented:
With #164646 I'm adding profiling and PGO support to our source build, which will obsolete this PR.

S-NA commented:
Closed in favor of the above mentioned PR. Thanks!
